File: backend/src/app/modules/graph/application/schema_meta.py
"""Reader for the `x-menagerist` metadata namespace in attribute schemas.

Standard JSON Schema keywords describe validation; `x-menagerist` describes
everything else (layout, advisory `required`, archived fields, display hints).
Nothing else in the backend reads `x-*` keywords directly.
"""

from typing import Any

from app.modules.graph.domain.errors import InvalidSchemaError

NAMESPACE = "x-menagerist"

_PROPERTY_BOOL_MEMBERS = ("archived", "search", "suggest")
_PROPERTY_STR_MEMBERS = ("kind", "display")
MAX_CARD_HIGHLIGHTS = 3


def _meta(node: object) -> dict[str, Any]:
    """Return the `x-menagerist` object of a schema or property, or `{}`."""
    if not isinstance(node, dict):
        return {}
    meta = node.get(NAMESPACE)
    return meta if isinstance(meta, dict) else {}


def required_keys(schema: dict[str, Any]) -> list[str]:
    """Return the keys marked required. Advisory only: never validated."""
    required = _meta(schema).get("required")
    if not isinstance(required, list):
        return []
    return [key for key in required if isinstance(key, str)]


def archived_keys(schema: dict[str, Any]) -> set[str]:
    """Return the keys of top-level properties marked archived."""
    properties = schema.get("properties")
    if not isinstance(properties, dict):
        return set()
    return {
        key for key, prop in properties.items() if _meta(prop).get("archived") is True
    }


def validation_schema(schema: dict[str, Any]) -> dict[str, Any]:
    """Return a copy of `schema` for validation.

    Drops the root `required` array and archived properties, so neither can
    reject a value. The stored schema is not modified.
    """
    archived = archived_keys(schema)
    stripped = {key: value for key, value in schema.items() if key != "required"}
    properties = schema.get("properties")
    if archived and isinstance(properties, dict):
        stripped["properties"] = {
            key: prop for key, prop in properties.items() if key not in archived
        }
    return stripped


def _check_property_meta(path: str, prop: object) -> None:
    meta = prop.get(NAMESPACE) if isinstance(prop, dict) else None
    if meta is None:
        return
    if not isinstance(meta, dict):
        raise InvalidSchemaError(f"{path}/{NAMESPACE} must be an object")
    for member in _PROPERTY_STR_MEMBERS:
        if member in meta and not isinstance(meta[member], str):
            raise InvalidSchemaError(f"{path}/{NAMESPACE}/{member} must be a string")
    for member in _PROPERTY_BOOL_MEMBERS:
        if member in meta and not isinstance(meta[member], bool):
            raise InvalidSchemaError(f"{path}/{NAMESPACE}/{member} must be a boolean")
    if "config" in meta and not isinstance(meta["config"], dict):
        raise InvalidSchemaError(f"{path}/{NAMESPACE}/config must be an object")


def _check_root_meta(root: object) -> None:
    if root is None:
        return
    if not isinstance(root, dict):
        raise InvalidSchemaError(f"{NAMESPACE} must be an object")
    version = root.get("version")
    if "version" in root and (
        isinstance(version, bool) or not isinstance(version, int)
    ):
        raise InvalidSchemaError(f"{NAMESPACE}/version must be an integer")
    for member in ("layout", "required"):
        if member in root and not isinstance(root[member], list):
            raise InvalidSchemaError(f"{NAMESPACE}/{member} must be an array")
    if not all(isinstance(key, str) for key in root.get("required", [])):
        raise InvalidSchemaError(f"{NAMESPACE}/required must contain strings")


def _check_highlight_entry(
    entry: object, props: dict[str, Any], seen: set[str]
) -> None:
    key = entry.get("key") if isinstance(entry, dict) else None
    if not isinstance(key, str):
        raise InvalidSchemaError(
            f"{NAMESPACE}/highlights/card entries need a string key"
        )
    if key in seen:
        raise InvalidSchemaError(f"{NAMESPACE}/highlights/card repeats {key!r}")
    seen.add(key)
    if key not in props:
        raise InvalidSchemaError(
            f"{NAMESPACE}/highlights/card refers to unknown field {key!r}"
        )
    if _meta(props[key]).get("archived") is True:
        raise InvalidSchemaError(
            f"{NAMESPACE}/highlights/card refers to archived field {key!r}"
        )


def _check_highlights(root: object, properties: object) -> None:
    """Check `highlights.card`: at most 3 unique keys of existing, live properties."""
    highlights = root.get("highlights") if isinstance(root, dict) else None
    if highlights is None:
        return
    if not isinstance(highlights, dict):
        raise InvalidSchemaError(f"{NAMESPACE}/highlights must be an object")
    card = highlights.get("card")
    if card is None:
        return
    if not isinstance(card, list):
        raise InvalidSchemaError(f"{NAMESPACE}/highlights/card must be an array")
    if len(card) > MAX_CARD_HIGHLIGHTS:
        raise InvalidSchemaError(
            f"{NAMESPACE}/highlights/card allows at most {MAX_CARD_HIGHLIGHTS} fields"
        )
    props = properties if isinstance(properties, dict) else {}
    seen: set[str] = set()
    for entry in card:
        _check_highlight_entry(entry, props, seen)


def check_meta_shape(schema: dict[str, Any]) -> None:
    """Check that known `x-menagerist` members are correctly typed.

    Unknown members are allowed so newer clients stay forward compatible.

    :raises InvalidSchemaError: when a known member has the wrong type or a
        highlight refers to a missing or archived field.
    """
    _check_root_meta(schema.get(NAMESPACE))
    properties = schema.get("properties")
    _check_highlights(schema.get(NAMESPACE), properties)
    if not isinstance(properties, dict):
        return
    for key, prop in properties.items():
        _check_property_meta(f"/properties/{key}", prop)
        items = prop.get("items") if isinstance(prop, dict) else None
        sub_properties = items.get("properties") if isinstance(items, dict) else None
        if isinstance(sub_properties, dict):
            for sub_key, sub_prop in sub_properties.items():
                path = f"/properties/{key}/items/properties/{sub_key}"
                _check_property_meta(path, sub_prop)

File: frontend/src/lib/schema-meta.ts
import type { XLayout } from '$lib/layout';

/**
 * Accessors for the `x-menagerist` namespace. Standard JSON Schema keywords describe
 * validation; this namespace describes everything else. Nothing else reads `x-*` keys.
 */
export const NAMESPACE = 'x-menagerist';
export const META_VERSION = 1;

export type PropertyMeta = {
	kind?: string;
	display?: string;
	archived?: boolean;
	search?: boolean;
	suggest?: boolean;
	config?: Record<string, unknown>;
	/** A `group` (table) field's sub-property keys in order. See `field-types/group/columns.ts`. */
	columns?: string[];
	/** Unknown members are preserved on round-trip. */
	[member: string]: unknown;
};

export type HighlightEntry = { key: string };

/** Which highlight list: an item type's cards, or a relationship type's connection rows. */
export type HighlightList = 'card' | 'connection';

export type SchemaMeta = {
	version?: number;
	layout?: XLayout;
	required?: string[];
	/** Fields shown on cards (`card`); other lists (for example `connection`) are preserved. */
	highlights?: {
		card?: HighlightEntry[];
		connection?: HighlightEntry[];
		[list: string]: unknown;
	};
	/** Unknown members are preserved on round-trip. */
	[member: string]: unknown;
};

function readMeta<T extends object>(node: unknown): T {
	if (typeof node !== 'object' || node === null) return {} as T;
	const meta = (node as Record<string, unknown>)[NAMESPACE];
	if (typeof meta !== 'object' || meta === null || Array.isArray(meta)) return {} as T;
	return { ...meta } as T;
}

/** Read a schema's root metadata. Returns `{}` when absent or malformed. */
export function readSchemaMeta(schema: object | null | undefined): SchemaMeta {
	return readMeta<SchemaMeta>(schema);
}

/** Read a property's metadata. Returns `{}` when absent or malformed. */
export function readPropMeta(prop: object | null | undefined): PropertyMeta {
	return readMeta<PropertyMeta>(prop);
}

/** Return a copy of `schema` with `patch` merged into its root metadata. */
export function withSchemaMeta<T extends object>(schema: T, patch: SchemaMeta): T {
	return { ...schema, [NAMESPACE]: { ...readSchemaMeta(schema), ...patch } };
}

/** Return a copy of `prop` with `patch` merged into its metadata. */
export function withPropMeta<T extends object>(prop: T, patch: PropertyMeta): T {
	return { ...prop, [NAMESPACE]: { ...readPropMeta(prop), ...patch } };
}

/** Return a copy of `prop` with its metadata replaced wholesale, not merged. */
export function replacePropMeta<T extends object>(prop: T, meta: PropertyMeta): T {
	const copy = { ...prop } as Record<string, unknown>;
	if (Object.keys(meta).length > 0) copy[NAMESPACE] = meta;
	else delete copy[NAMESPACE];
	return copy as T;
}

/** A stored highlight list, in order. Malformed entries are skipped. */
export function readHighlights(
	schema: object | null | undefined,
	list: HighlightList = 'card'
): HighlightEntry[] {
	const entries = readSchemaMeta(schema).highlights?.[list];
	if (!Array.isArray(entries)) return [];
	return entries.flatMap((e) =>
		typeof e === 'object' && e !== null && typeof (e as HighlightEntry).key === 'string'
			? [{ key: (e as HighlightEntry).key }]
			: []
	);
}

/** Return a copy of `schema` with one highlight list replaced. Other lists are kept. */
export function withHighlights<T extends object>(
	schema: T,
	entries: HighlightEntry[],
	list: HighlightList = 'card'
): T {
	const { [list]: _previous, ...others } = readSchemaMeta(schema).highlights ?? {};
	const highlights = entries.length > 0 ? { ...others, [list]: entries } : others;
	const meta = { ...readSchemaMeta(schema) };
	if (Object.keys(highlights).length > 0) meta.highlights = highlights;
	else delete meta.highlights;
	return { ...schema, [NAMESPACE]: meta };
}

/** Keys of top-level properties marked archived: hidden from forms, data kept. */
export function archivedKeys(schema: { properties?: Record<string, unknown> } | null): Set<string> {
	return new Set(
		Object.entries(schema?.properties ?? {})
			.filter(([, prop]) => readPropMeta(prop as object).archived === true)
			.map(([key]) => key)
	);
}

/**
 * Return a copy of `schema` for client-side validation. A standard root `required` array
 * (for example from a schema authored through the API) and archived properties are dropped:
 * required is advisory and archived fields are hidden, so neither may produce a blocking
 * error. Mirrors the backend's `validation_schema`.
 */
export function validationSchema<T extends object>(schema: T): T {
	const copy: Record<string, unknown> = { ...(schema as Record<string, unknown>) };
	delete copy.required;
	const properties = copy.properties as Record<string, unknown> | undefined;
	if (properties) {
		const archived = archivedKeys({ properties });
		copy.properties = Object.fromEntries(
			Object.entries(properties).filter(([k]) => !archived.has(k))
		);
	}
	return copy as T;
}
